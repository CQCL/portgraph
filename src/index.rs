pub trait EntityIndex: Copy + Eq + Default {
    fn new(index: usize) -> Self {
        Self::try_new(index).unwrap()
    }

    fn try_new(index: usize) -> Option<Self>;
    fn index(self) -> usize;
}

#[macro_export]
macro_rules! make_entity {
    ($($(#[$attr:meta])* $vis:vis struct $name:ident(u32);)*) => { $(
        #[repr(transparent)]
        #[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash)]
        $(#[$attr:meta])*
        $vis struct $name(::std::num::NonZeroU32);

        impl $crate::index::EntityIndex for $name {
            fn try_new(index: usize) -> Option<Self> {
                if index < u32::MAX as usize {
                    let index = index.wrapping_add(1) as u32;
                    Some(Self(unsafe {
                        ::std::num::NonZeroU32::new_unchecked(index)
                    }))
                } else {
                    None
                }
            }

            #[inline(always)]
            fn index(self) -> usize {
                u32::from(self.0).wrapping_sub(1) as usize
            }
        }

        impl ::std::default::Default for $name {
            fn default() -> Self {
                Self(::std::num::NonZeroU32::new(u32::MAX).unwrap())
            }
        }

    )*};
}

macro_rules! int_entity_impl {
    ($entity:ident) => {
        impl $crate::index::EntityIndex for $entity {
            #[inline(always)]
            fn try_new(ix: usize) -> Option<Self> {
                if (ix <= <$entity>::MAX as usize || <$entity>::BITS > usize::BITS) {
                    Some(ix as $entity)
                } else {
                    None
                }
            }

            #[inline(always)]
            fn index(self) -> usize {
                self as usize
            }
        }
    };
}

int_entity_impl!(u128);
int_entity_impl!(u64);
int_entity_impl!(u32);
int_entity_impl!(u16);
int_entity_impl!(u8);

make_entity! {
    pub struct NodeIndex(u32);
    pub struct PortIndex(u32);
}
