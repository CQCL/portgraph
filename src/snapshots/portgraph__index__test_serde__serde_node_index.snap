---
source: src/index.rs
expression: "serde_json::to_string_pretty(&graph).unwrap()"
---
{
  "graph": {
    "node_meta": [
      {
        "n": {
          "first_port": 0,
          "incoming": 5,
          "outgoing": 4,
          "capacity": 8
        }
      },
      {
        "n": {
          "first_port": 8,
          "incoming": 5,
          "outgoing": 4,
          "capacity": 8
        }
      },
      {
        "n": {
          "first_port": 16,
          "incoming": 5,
          "outgoing": 4,
          "capacity": 8
        }
      },
      {
        "n": {
          "first_port": 24,
          "incoming": 5,
          "outgoing": 4,
          "capacity": 8
        }
      },
      {
        "n": {
          "first_port": 32,
          "incoming": 5,
          "outgoing": 4,
          "capacity": 8
        }
      }
    ],
    "port_link": [
      null,
      null,
      null,
      null,
      8,
      null,
      null,
      null,
      4,
      null,
      null,
      null,
      16,
      null,
      null,
      null,
      12,
      null,
      null,
      null,
      24,
      null,
      null,
      null,
      20,
      null,
      null,
      null,
      32,
      null,
      null,
      null,
      28,
      null,
      null,
      null,
      null,
      null,
      null,
      null
    ],
    "port_meta": [
      1,
      1,
      1,
      1,
      2147483649,
      2147483649,
      2147483649,
      2147483649,
      2,
      2,
      2,
      2,
      2147483650,
      2147483650,
      2147483650,
      2147483650,
      3,
      3,
      3,
      3,
      2147483651,
      2147483651,
      2147483651,
      2147483651,
      4,
      4,
      4,
      4,
      2147483652,
      2147483652,
      2147483652,
      2147483652,
      5,
      5,
      5,
      5,
      2147483653,
      2147483653,
      2147483653,
      2147483653
    ],
    "node_free": null,
    "port_free": [],
    "node_count": 5,
    "port_count": 40,
    "link_count": 4,
    "_marker": null
  },
  "multiport": {
    "order": "bitvec::order::Lsb0",
    "head": {
      "width": 64,
      "index": 0
    },
    "bits": 0,
    "data": []
  },
  "copy_node": {
    "order": "bitvec::order::Lsb0",
    "head": {
      "width": 64,
      "index": 0
    },
    "bits": 0,
    "data": []
  },
  "copy_node_count": 0,
  "subport_count": 0
}
